upstream-seed/src/peer.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright © 2021 The Radicle Upstream Contributors
+//
+// This file is part of radicle-upstream, distributed under the GPLv3
+// with Radicle Linking Exception. For full terms see the included
+// LICENSE file.
+
+use std::{net::SocketAddr, time::Duration};
+
+use anyhow::Context as _;
+use futures::stream::StreamExt as _;
+
+use librad::{
+    git::{identities, replication, storage::fetcher, tracking},
+    net::{
+        discovery::{self, Discovery as _},
+        protocol, Network,
+    },
+    paths::Paths,
+    PeerId,
+};
+use link_identities::git::Urn;
+
+type LibradPeer = librad::net::peer::Peer<librad::SecretKey>;
+type PeerInfo = librad::net::peer::PeerInfo<SocketAddr>;
+
+/// Configuration for creating a new [`Peer`].
+#[derive(Clone)]
+pub struct Config {
+    pub rad_paths: Paths,
+    pub key: librad::SecretKey,
+    pub listen: SocketAddr,
+}
+const PROVIDER_QUERY_TIMEOUT: Duration = Duration::from_secs(60);
+
+/// Wrapper around [`librad::net::peer::Peer`] that provides seed specific functionality.
+#[derive(Clone)]
+pub struct Peer {
+    librad_peer: LibradPeer,
+}
+
+impl Peer {
+    /// Create a new client.
+    pub fn new(config: Config) -> Self {
+        let storage = librad::net::peer::config::Storage {
+            protocol: librad::net::peer::config::ProtocolStorage {
+                fetch_slot_wait_timeout: Default::default(),
+                pool_size: 4,
+            },
+            user: librad::net::peer::config::UserStorage { pool_size: 4 },
+        };
+
+        let peer_config = librad::net::peer::Config {
+            signer: config.key,
+            protocol: protocol::Config {
+                paths: config.rad_paths,
+                listen_addr: config.listen,
+                advertised_addrs: None, // TODO: Should we use this?
+                membership: Default::default(),
+                network: Network::Main,
+                replication: replication::Config::default(),
+                fetch: Default::default(),
+                rate_limits: Default::default(),
+            },
+            storage,
+        };
+        let librad_peer = LibradPeer::new(peer_config).expect("failed to create peer");
+
+        Self { librad_peer }
+    }
+
+    /// Run the peer. This function runs indefinitely until a fatal error occurs.
+    pub async fn run(self, bootstrap: Vec<(PeerId, SocketAddr)>) {
+        let disco = discovery::Static::resolve(bootstrap).unwrap();
+
+        loop {
+            match self.librad_peer.bind().await {
+                Ok(bound) => {
+                    let (_kill, run) = bound.accept(disco.clone().discover());
```

Comment:
What does disco do? Couldn't find any info on it.

Reply:
`disco` tells the protocol how to connect to peers. In this case it is simply used to tell the boostrap peers to the protocol. In more advanced cases disco could repeatedly provide new peers that it discovers on the LAN for example.